### ISSUE TITLE
Question about auto-completion as we type

### ISSUE THREAD

**Issue by stefanos82** [fatih/vim-go#3682]
I have noticed a rather peculiar behavior with auto-completion and I need your advice.

My settings are as follows:

```vim
" For Go language settings
autocmd filetype go inoremap <buffer> . .<C-x><C-o>

inoremap <C-@> <C-x><C-o>

let g:go_highlight_fields = 1
let g:go_highlight_build_constraints = 1
let g:go_highlight_types = 1
let g:go_highlight_extra_types = 1
let g:go_highlight_format_strings = 1
let g:go_highlight_functions = 1
let g:go_highlight_function_calls = 1
let g:go_highlight_function_parameters = 1
let g:go_highlight_operators = 1
let g:go_highlight_variable_declarations = 1

let g:go_code_completion_icase = 1
```

As soon as I press the dot, of course auto-completion gets triggered, but here's the current "issue" I'm facing: I just type all in lowercase and instead of highlighting the matching keywords, for instance `fprintf` should match with `Fprintf`, but it just keeps the drop down menu open, until I either move my arrows left and right or press `Esc` to close it.

Another related issue I face with drop down menu is with the use of `tab` key...

```vim
 Use Tab to select completion downwards when popup window opened
inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"
" Use Tab to select completion upwards when popup window opened
inoremap <expr> <S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<cr>"
```

These aforementioned settings work with every other language, but not with Go...

My go version is `go version go1.22.6 linux/amd64`

Any feedback please?

**Comment by bhcleek** [comment]
What does `set completeopt?` output for you in a Go file vs filetypes where you're seeing the behavior your want?

**Comment by stefanos82** [comment]
In all files, go included, I see `completeopt=menu,preview,menuone,preview,longest`.

**Comment by bhcleek** [comment]
🤔 As far as I know, vim-go isn't doing anything unusual here and the only reason for the kinds of differences you're describing would be the value of `completeopt`. With `completeopt` set to `menu,preview,longest` I see the behavior you describe. With `menu,preview,noinsert`, I seem to see the behavior you want.

**Comment by stefanos82** [comment]
Let me double-check my settings, just to be sure everything is configured correctly.

**Comment by stefanos82** [comment]
OK, your suggestion with `noinsert` works; how about the use of `Tab` key? Why it does not work only with `vim-go`? Only `Ctrl-n` / `Ctrl-p` work, but it should have worked with `Tab` / `shift-Tab` too...

**Comment by bhcleek** [comment]
Vim-go doesn't do anything to remap tab. If I were you, I'd check it with `:map <TAB>` after loading a go file.

**Comment by stefanos82** [comment]
You mean `:verbose imap <TAB>`, right?

**Comment by bhcleek** [comment]
> You mean :verbose imap <TAB>, right?

Yes.

```
inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"
" Use Tab to select completion upwards when popup window opened
inoremap <expr> <S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<cr>"
```

is definitely not
```
i  <Tab>       * coc#pum#visible() ? coc#pum#next(1) : "\<Tab>"                                  
        Last set from ~/.vim/pack/plugins/start/coc.nvim/autoload/coc/ui.vim line 23 
```

> As you can see, I use coc.nvim; I hope there's no conflict somewhere behind the scenes...

I don't know how coc presents its code completion, but based on the behaviors you're seeing, it looks to me like you're using coc and vim-go for Go, which is fine, but please understand that this may mean some small differences in the user experience. Are you intending to use coc for Go? Do you use it for other languages, too? If I had to guess, I'd bet that your `<TAB>` mapping is happening before coc is loaded, and then coc overrides it. Given that it appears the code completion for vim-go is happening using omnicomplete (as evidenced by how `completeopt` affected the experience for you), it seems that vim-go is likely being loaded _after_ coc.

There are a few different ways to resolve these problems, but the right answer depends on how your particular setup is configured and whether you want to continue using coc for Go. I don't use coc and I'm not an expert in its use, but I'm willing to help you figure out exactly what you need to do to resolve this if you want.

Please start by reading about how initialization occurs by reviewing `:help rtp`, `:help initialiazation`, `:help load-plugins`, and `:help after-directory`.

**Comment by stefanos82** [comment]
> ```
> inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"
> " Use Tab to select completion upwards when popup window opened
> inoremap <expr> <S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
> inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<cr>"
> ```
> 
> is definitely not
> 
> ```
> i  <Tab>       * coc#pum#visible() ? coc#pum#next(1) : "\<Tab>"                                  
>         Last set from ~/.vim/pack/plugins/start/coc.nvim/autoload/coc/ui.vim line 23 
> ```

Based on their code which I took a brief look, they set and unset mappings deep inside their implementation, and the two versions work just fine with all other languages, besides `vim-go`.

> I don't know how coc presents its code completion, but based on the behaviors you're seeing, it looks to me like you're using coc and vim-go for Go, which is fine, but please understand that this may mean some small differences in the user experience. Are you intending to use coc for Go? 

No, I do not. They have separate settings in different configuration files of mine, that is `vim-go` settings and `coc.nvim`.

> Do you use it for other languages, too? If I had to guess, I'd bet that your `<TAB>` mapping is happening before coc is loaded, and then coc overrides it. Given that it appears the code completion for vim-go is happening using omnicomplete (as evidenced by how `completeopt` affected the experience for you), it seems that vim-go is likely being loaded _after_ coc.
> 

As I have stated above, I use `coc.nvim` for all languages, but not for Go; for Go, I use `vim-go` exclusively.

> There are a few different ways to resolve these problems, but the right answer depends on how your particular setup is configured and whether you want to continue using coc for Go. I don't use coc and I'm not an expert in its use, but I'm willing to help you figure out exactly what you need to do to resolve this if you want.
> 
> Please start by reading about how initialization occurs by reviewing `:help rtp`, `:help initialiazation`, `:help load-plugins`, and `:help after-directory`.

By echoing `&rtp`, I can see `vim-go` getting loaded first and then at the near end `coc.nvim` gets loaded.

Does this help somehow?

**Comment by bhcleek** [comment]
> No, I do not. They have separate settings in different configuration files of mine, that is vim-go settings and coc.nvim.
> By echoing &rtp, I can see vim-go getting loaded first and then at the near end coc.nvim gets loaded.

I suspect you have the `<TAB>` mapping in your vimrc. Is that right? If so, you'll need to move it to one of the after file. Given that you're not actually using it for other languages, because you're actually using coc for those, I'd recommend either putting it in `ftplugin/go.vim` in your vim config directory or `after/plugins/vim-go.vim`. The effects will be different depending on which file you choose to move it to.

**Comment by stefanos82** [comment]
> I suspect you have the `<TAB>` mapping in your vimrc.

In my `$HOME/.vimrc` file I have my custom files:

```vim
" Custom settings
source $HOME/.vim/configurations/encoding.vim
source $HOME/.vim/configurations/functions.vim
source $HOME/.vim/configurations/sets.vim
source $HOME/.vim/configurations/maps.vim
source $HOME/.vim/configurations/cocnvim.vim
source $HOME/.vim/configurations/golang.vim
```

The `<TAB>` mapping is inside `map.vim`:

```vim
" Use Tab to select completion downwards when popup window opened
inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"
" Use Tab to select completion upwards when popup window opened
inoremap <expr> <S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<cr>"
```

That's it, really...

> Is that right? If so, you'll need to move it to one of the `after` files. Given that you're not actually using it for other languages, because you're actually using coc for those, I'd recommend either putting it in `ftplugin/go.vim` in your vim config directory or `after/plugins/vim-go.vim`. The effects will be different depending on which file you choose to move it to.

The mapping is being used by coc too, because I have disabled it and tab key / shift-tab key stops working afterwards; as soon as I uncomment the aforementioned code, everything comes back to normal; that's why I said, it's a rather peculiar behavior!

**Comment by stefanos82** [comment]
I have found the problem...darn it `coc.nvim`! It causes a conflict between my mapping that works with `vim-go` and `coc.vim` :sob: 

I have disabled mappings with the following line `let g:coc_disable_mappings_check = 1` and I can clearly see `coc.vim` not working, almost at all, but `vim-go` works as expected!

Do you know of a way to tell `vim-go` "please check whether `coc.vim` has mappings enabled and if yes, disable them on the fly as long as a Go file is open"?

**Comment by bhcleek** [comment]
https://github.com/neoclide/coc.nvim/issues/349 provides some advice that may help you.

**Comment by stefanos82** [comment]
Things are getting weirder and weirder on my side...with my tab mapping completely disabled and having `coc.nvim` enabled, I get `no mapping found` for all languages; as soon as I restore my own mapping settings, I get 

```
Kitty keyboard protocol: Cleared                                                                 
i  <Tab>       * coc#pum#visible() ? coc#pum#next(1) : "\<Tab>"                                  
        Last set from ~/.vim/pack/plugins/start/coc.nvim/autoload/coc/ui.vim line 23 
```

This has to be the most bizarre behavior I have ever witnessed, seriously!

**Comment by stefanos82** [comment]
Let's see whether they will provide further feedback with my current situation https://github.com/neoclide/coc.nvim/discussions/5133

**Comment by stefanos82** [comment]
Anyway, I will close this ticket and will provide further info if and when they reply, assuming there is an actual solution to this kind of peculiarity.

My apologies for the noise I caused, I had no bad intentions :disappointed: I thought it would be something easy to fix...

**Comment by bhcleek** [comment]
> My apologies for the noise I caused, I had no bad intentions 😞 I thought it would be something easy to fix...

Please don't feel bad; I know you had no bad intentions, and I don't mind helping.

**Comment by stefanos82** [comment]
Seems like we ended up in a dead-end circle; it led nowhere in the aforementioned discussion.

I don't know whether it was lack of communication skills or whatsoever, but I was repeating the same things over and over and over again, leading nowhere.

Oh well...I will continue using it as I already was and everything will be alright.

**Comment by bhcleek** [comment]
Have you put the mapping you want to use with vim-go into `after/ftplugin/go.vim`?

**Comment by stefanos82** [comment]
I have tried; it didn't work.

It's OK; I will continue with the options I have with the arrow keys.

**Comment by stefanos82** [comment]
Update: Seems like when you are tired and you are forcing yourself for a solution, you do yourself more harm than good.

I think I have figured out the solution; so far it works as expected and I feel so stupid for not understanding this logic in the first place :man_facepalming: 

In `coc.nvim` configuration settings, all I had to add was the following:

```json
        "golang": {
            "command": "gopls",
            "rootPatterns": ["go.mod", ".git/"],
            "filetypes": ["go"]
        }
```

This way, I use `coc.nvim` for the auto-completion, without the need for my custom mapping with `<tab>` key.

Other than that, `vim-go` works as expected, apart from its own code completion which I don't use explicitly.
